### PR TITLE
Changing the testing/development environment to use Docker-compose instead of DBDeployer

### PR DESCRIPTION
**Pull request by EtienneBerubeShopify**
*NOTE:* This PR is a cherry-pick of some of the commits for this [issue](https://github.com/issues/assigned). It was becoming too big and the PR could be split in 2 self-contained PRs as they encompass different things.

This PR changes the way the testing/development environment is setup. Initially, the environment was created using DBDeployer to create 2 databases with a writer-reader topology. However, as LHM now becomes "aware" of ProxySQL, it is important to have a ProxySQL instance running during development. However, the binaries are not available for MacOS (or would need to be re-compiled), but there is a docker image available. This justifies the move from DBDeployer to Docker-compose as it would allow to have the required ProxySQL instance up and running.

### What changed?
- `dev.yml`: Changed some commands to take into account the environments 
- `docker-compose.yml`: The definition for environment
- `scripts/**/*`: Added scripts and configs for the MySQL and ProxySQL 
- `spec/integration/database.yml`: The file is now static (before it would be generated during the dbdeployer install)
### In the future?
If someone wants to add a service for the development environment, the only thing to change woluld be the `met?` condition for the docker-compose step in `dev up`. Right now, `met?` runs a regex against the number of service running. So if a service is added or removed, the number in the regex should change a well.

### To check in this PR
- Mostly edge cases. There are ways that these changes could break some features without me or the tests noticing.
- The replication was usually handled by dbdeployer. Now it is done within the scripts in `scripts/mysql/*`. Considering I have very limited knowledge on MySQL replication, it would be great to have someone validated that it is done properly.

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+      SERVER_ID: 2
+      MYSQL_HOST: mysql-2
+    ports:
+      - "33007:3306"
+    depends_on:
+      - mysql-1
+#  Proxysql
+  proxysql:
+    container_name: proxysql
+    image: proxysql/proxysql
```
[the remark this answers is not in the document]
Comment (EtienneBerubeShopify) [reply docker-compose.yml:35]: I can lock on the latest stable release. I did run the test suite with @latest and it worked

File: docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,43 @@
+services:
+#  Writer
+  mysql-1:
+    container_name: mysql-1
+    image: mysql:5.7.18
```
[the remark this answers is not in the document]
Comment (EtienneBerubeShopify) [reply docker-compose.yml:5]: I got these configs from a different repo that was using docker-compose as a test environment with a similar setup. Would there be some resources where I could find a different MySQL or Percona version that might be more relevant?

[the remark this answers is not in the document]
Comment (shuhaowu) [reply docker-compose.yml:5]: Ghostferry's repo would be a start. That said, I would not blatantly copy and paste. At least try to look through the MySQL documentations (and other relevant documentations) on what the options does.

File: scripts/mysql/reader/reader.cnf [hunk 3]
```diff
@@ -0,0 +1,17 @@
+[mysqld]
+socket=/var/lib/mysql/mysql.sock
+user=mysql
+read_only=ON
```
[the remark this answers is not in the document]
Comment (EtienneBerubeShopify) [reply scripts/mysql/reader/reader.cnf:4]: I copied the way Shopify/ghostferry does it as I like having *most* of the configs at the same place (i.e. removed the `.cnf` files for mysql)

File: scripts/mysql/reader/create_replication.sql [hunk 4]
```diff
@@ -0,0 +1,10 @@
+STOP SLAVE;
+CHANGE MASTER TO
+MASTER_HOST='mysql-1',
+MASTER_AUTO_POSITION=1,
+MASTER_USER='replication',
+MASTER_PASSWORD='password';
+
+start slave;
```
[the remark this answers is not in the document]
Comment (EtienneBerubeShopify) [reply scripts/mysql/reader/create_replication.sql:8]: Good catch! After investigation, it was found that if a race condition were to happen (replication starts before user are created on writer), the slave would retry connection. I added a custom retry time so that it doesn't block the rest of the tests (retries every second for 5 minutes before aborting).

File: scripts/mysql/writer/writer.cnf [hunk 5]
```diff
@@ -0,0 +1,16 @@
+[mysqld]
+socket=/var/lib/mysql/mysql.sock
+user=mysql
+log_bin=/var/lib/mysql/mysql-bin
+relay-log=/var/log/mysql/mysql-relay-bin.log
+binlog_format=row
+binlog_row_image=minimal
```
[the remark this answers is not in the document]
Comment (EtienneBerubeShopify) [reply scripts/mysql/writer/writer.cnf:7]: marked as resolved since the `.cnf` were removed.

File: scripts/proxysql/proxysql.cnf [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+# defines all the MySQL users
+mysql_users:
+(
+  {
+    username = "root"
+    password = "password"
+    default_hostgroup = 0
+    max_connections=1000
+    active = 1
+  },
```
[the remark this answers is not in the document]
Comment (EtienneBerubeShopify) [reply scripts/proxysql/proxysql.cnf:55]: I kept it as there are some proxysql tests that depend on it (to be added in the LHM-retry pr). Moreover, I like the idea of having it while having a distinction between root and writer. This can be removed if it is found to be too redundant.

File: spec/integration/chunker_spec.rb [hunk 7]
```diff
@@ -169,6 +171,8 @@ def log_messages
     it 'should copy 23 rows from origin to destination in one shot with slave lag based throttler, regardless of the value of the id' do
       23.times { |n| execute("insert into origin set id = '#{ 100000 + n * n + 23 }'") }
 
+      ActiveRecord::Base.stubs(:connection_pool).returns(stub(spec: stub(config: @active_record_config)))
```
[the remark this answers is not in the document]
Comment (shuhaowu) [reply spec/integration/chunker_spec.rb:174]: Wait if you didn't find a root cause how did you solve the problem? Is this simply a workaround? Preferably we should understand what we are doing and what side effects it may have before writing the code to workaround the problem.

[the remark this answers is not in the document]
Comment (EtienneBerubeShopify) [reply spec/integration/chunker_spec.rb]: I investigated and found that the error was being received from a MySQL2 client being used earlier in development. However, in a later change the main connection was changed to ActiveRecord, which doesn't need to stub `#connection_pool`.

File: docker-compose.yml [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    hostname: 'mysql-2'
+    volumes:
+      - ./scripts/mysql/reader:/docker-entrypoint-initdb.d
+    environment:
+      MYSQL_ROOT_PASSWORD: password
+      MYSQL_HOST: mysql-2
+    ports:
+      - "33007:3306"
+    depends_on:
+      - mysql-1
```
[the remark this answers is not in the document]
Comment (EtienneBerubeShopify) [reply docker-compose.yml]: I removed the "depends_on" and added the wait script in the dev up pipeline. Now everything is consistent and beautiful.
